media-proxy/src/mesh/conn_rdma.cc

```
@@ -0,0 +1,359 @@
+        return Result::error_initialization_failed;
+    }
+    if(queue_size == 0) {
+        log::error("Queue size is not set")("queue_size", queue_size);
+        set_state(ctx, State::closed);
+        return Result::error_bad_argument;
+    }
+
+    res = init_queue_with_elements(queue_size, trx_sz);
+    if (res != Result::success) {
+        log::error("Failed to initialize buffer queue")("trx_sz", trx_sz);
+        if (ep_ctx) {
+            libfabric_ep_ops.ep_destroy(&ep_ctx);
+        }
+        set_state(ctx, State::closed);
+        return res;
+    }
+
+    // Configure RDMA endpoint
+    res = configure_endpoint(ctx);
+    if (res != Result::success) {
+        log::error("RDMA configuring failed")("state", "closed");
+        if (ep_ctx) {
+            libfabric_ep_ops.ep_destroy(&ep_ctx);
+        }
+        set_state(ctx, State::closed);
+        return res;
+    }
+
+    init = true;
+    res = start_threads(ctx);
+    set_state(ctx, State::active);
+    return Result::success;
+}
+
+/**
+ * @brief Cleans up RDMA resources and resets the state.
+ *
+ * This function destroys the RDMA endpoint context if it exists, cleans up the buffer queue,
+ * and marks the RDMA as uninitialized. It ensures that all allocated resources are properly
+ * released and the state is reset to allow for reinitialization if needed.
+ *
+ * @param ctx The context used for the cleanup operation.
+ * @return Result indicating the success or failure of the cleanup operation.
+ */
+Result Rdma::on_shutdown(context::Context& ctx)
+{
+    // Note: this is a blocking call
+    
+    // Cancel `rdma_cq_thread` context
+    rdma_cq_thread_ctx.cancel();
+    // Cancel `process_buffers_thread` context
+    process_buffers_thread_ctx.cancel();
+    // Ensure buffer-related threads are unblocked
+    notify_buf_available();
+
+    // Check if `rdma_cq_thread` has stopped
+    if (handle_rdma_cq_thread.joinable()) {
+        handle_rdma_cq_thread.join();
```

join() can throw an exception. low chances for that, but I think we wanted to catch all exceptions and return errors by return value.

---

I suggest to mark all methods as `noexcept`

---

Can be added in future PRs

---

fixed in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/286/commits/390a65332407a21263d7269b722310576780ad64